File: homework3/src/main/java/by/popkov/homework3/ContactsActivity.java
package by.popkov.homework3;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.SearchView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

public class ContactsActivity extends AppCompatActivity {
    private static final int REQUEST_CODE_FOR_ADD = 7777;
    private static final int REQUEST_CODE_FOR_EDIT = 1111;
    private static final String ADAPTER = "adapter";

    private RecyclerView contactsRecyclerView;
    private ContactListAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_contacts);
        initContactsRecyclerView(savedInstanceState);
        setListeners();
        setToolBar();
    }

    private void initContactsRecyclerView(Bundle savedInstanceState) {
        contactsRecyclerView = findViewById(R.id.recyclerViewContacts);
        if (savedInstanceState != null) {
            contactsRecyclerView.setAdapter((ContactListAdapter) savedInstanceState
                    .getParcelable(ADAPTER));
        } else {
            contactsRecyclerView.setAdapter(new ContactListAdapter());
        }
        contactsRecyclerView.setLayoutManager(new GridLayoutManager(ContactsActivity.this,
                Integer.parseInt(contactsRecyclerView.getTag().toString())));
        adapter = (ContactListAdapter) contactsRecyclerView.getAdapter();

        if (adapter != null) {
            visibleSwitcher(adapter.getFullItemCount());
        }
    }

    private void setListeners() {
        setSearchViewListener();
        setFloatingActionButtonAddContactListener();
        setItemListenerWithData();
    }

    private void setSearchViewListener() {
        final TextView textViewHead = findViewById(R.id.textViewHead);
        SearchView searchView = findViewById(R.id.searchView);
        searchView.setOnSearchClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                textViewHead.setVisibility(View.INVISIBLE);
            }
        });
        searchView.setOnCloseListener(new SearchView.OnCloseListener() {
            @Override
            public boolean onClose() {
                textViewHead.setVisibility(View.VISIBLE);
                return false;
            }
        });
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                adapter.getFilter().filter(newText);
                return false;
            }
        });
    }

    private void setFloatingActionButtonAddContactListener() {
        FloatingActionButton floatingActionButtonAddContact = findViewById(R.id.floatingActionButtonAddContact);
        floatingActionButtonAddContact.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivityForResult(AddContactActivity
                        .newIntent(ContactsActivity.this), REQUEST_CODE_FOR_ADD);
            }
        });
    }

    private void setItemListenerWithData() {
        adapter.setItemListenerWithData(new ContactListAdapter.ItemListenerWithData() {
            @Override
            public void onClick(Contact oldContact) {
                startActivityForResult(
                        EditContactActivity.newIntent(ContactsActivity.this, oldContact),
                        REQUEST_CODE_FOR_EDIT
                );
            }
        });
    }

    private void setToolBar() {
        setSupportActionBar((Toolbar) findViewById(R.id.toolBar));
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(ADAPTER, adapter);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (data != null && resultCode == RESULT_OK) {
            if (REQUEST_CODE_FOR_ADD == requestCode) {
                Contact contact = (Contact) data.getSerializableExtra(AddContactActivity.EXTRA_CONTACT_FOR_ADD);
                if (contact != null) {
                    adapter.addContact(contact);
                }
            } else if (REQUEST_CODE_FOR_EDIT == requestCode) {
                Contact newContact = (Contact) data.getSerializableExtra(EditContactActivity.EXTRA_NEW_CONTACT);
                Contact oldContact = (Contact) data.getSerializableExtra(EditContactActivity.EXTRA_OLD_CONTACT);
                if (newContact != null) {
                    adapter.editContact(newContact);
                } else if (oldContact != null) {
                    adapter.removeContact(oldContact);
                }
            }
            visibleSwitcher(adapter.getFullItemCount());
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    private void visibleSwitcher(int fullItemCount) {
        if (fullItemCount > 0) {
            contactsRecyclerView.setVisibility(View.VISIBLE);
        } else {
            contactsRecyclerView.setVisibility(View.INVISIBLE);
        }

    }
}

File: homework4/src/main/java/by/popkov/homework4/ContactsActivity.kt
package by.popkov.homework4

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.SearchView
import android.widget.SearchView.OnQueryTextListener
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton

class ContactsActivity : AppCompatActivity() {
    private lateinit var contactsRecyclerView: RecyclerView
    private lateinit var adapter: ContactListAdapter

    private val requestCodeForAdd: Int = 7777
    private val requestCodeForEdit: Int = 1111

    companion object{
        const val ADAPTER = "adapter"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_contacts)
        initContactsRecyclerView(savedInstanceState)
        setListeners()
        setToolBar()

    }

    private fun initContactsRecyclerView(savedInstanceState: Bundle?) {
        contactsRecyclerView = findViewById(R.id.recyclerViewContacts)
        if (savedInstanceState != null) contactsRecyclerView.adapter = savedInstanceState
                .getParcelable<ContactListAdapter>(ADAPTER)
        else contactsRecyclerView.adapter = ContactListAdapter()
        contactsRecyclerView.layoutManager = GridLayoutManager((this@ContactsActivity),
                contactsRecyclerView.tag.toString().toInt(), RecyclerView.VERTICAL, false)
        adapter = contactsRecyclerView.adapter as ContactListAdapter
        visibleSwitcher(adapter.getFullItemCount())
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(ADAPTER, adapter)
    }

    private fun setToolBar() {
        setSupportActionBar(findViewById(R.id.toolBar))
    }

    private fun setListeners() {
        val textViewHead: TextView = findViewById(R.id.textViewHead)
        val searchView: SearchView = findViewById(R.id.searchView)
        searchView.setOnSearchClickListener { textViewHead.visibility = View.INVISIBLE }
        searchView.setOnCloseListener(object : SearchView.OnCloseListener {
            override fun onClose(): Boolean {
                textViewHead.visibility = View.VISIBLE
                return false
            }
        })
        searchView.setOnQueryTextListener(object : OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                adapter.filter.filter(newText)
                return false
            }
        })
        val floatingActionButtonAddContact: FloatingActionButton = findViewById(R.id.floatingActionButtonAddContact)
        floatingActionButtonAddContact.setOnClickListener {
            startActivityForResult(AddContactActivity
                    .newIntent(this@ContactsActivity), requestCodeForAdd)
        }
        adapter.setItemListenerWithData(object : ContactListAdapter.ItemListenerWithData {
            override fun onClick(oldContact: Contact, positionFullList: Int, positionList: Int) {
                startActivityForResult(EditContactActivity.newIntent(this@ContactsActivity,
                        oldContact, positionFullList, positionList), requestCodeForEdit)
            }
        })

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (data != null) {
            if (requestCodeForAdd == requestCode && resultCode == Activity.RESULT_OK) {
                val contact: Contact? = data.getSerializableExtra(AddContactActivity.PUT_EXTRA) as Contact
                if (contact != null) {
                    adapter.addContact(contact)
                }
            } else if (requestCodeForEdit == requestCode && resultCode == Activity.RESULT_OK) {
                val newContact: Contact? = data.getSerializableExtra(EditContactActivity.EXTRA_NEW_CONTACT) as? Contact
                val fullListPosition = data.getIntExtra(EditContactActivity.EXTRA_FULL_LIST_POS, -202)
                val listPosition = data.getIntExtra(EditContactActivity.EXTRA_LIST_POS, -204)
                if (newContact != null) {
                    adapter.editContact(newContact, fullListPosition, listPosition)
                } else {
                    adapter.removeContact(fullListPosition, listPosition)
                }
            }
            visibleSwitcher(adapter.getFullItemCount())
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun visibleSwitcher(fullItemCount: Int) {
        if (fullItemCount > 0) contactsRecyclerView.visibility = View.VISIBLE
        else contactsRecyclerView.visibility = View.INVISIBLE

    }
}